packages/@ember/application/instance.ts

```diff
@@ -120,15 +124,23 @@ const ApplicationInstance = EngineInstance.extend({
     this._booted = true;
 
     return this;
-  },
+  }
 
-  setupRegistry(options) {
-    this.constructor.setupRegistry(this.__registry__, options);
-  },
+  setupRegistry(options: BootOptions) {
+    (this.constructor as typeof ApplicationInstance).setupRegistry(this.__registry__, options);
+  }
+
+  _router?: Router;
 
-  router: computed(function () {
-    return this.lookup('router:main');
-  }).readOnly(),
+  get router() {
+    if (this._router) {
+      return this._router;
+    }
+
+    let router = this.lookup('router:main');
+    assert('expected an instance of Router', router instanceof Router);
+    return (this._router = router);
```

Comment:
In general @rwjblue and I both try to invert this pattern:

```suggestion
    if (!this._router) {
      let router = this.lookup('router:main');
      assert('expected an instance of Router', router instanceof Router);
      this._router = router;
    }

    return this._router;
```

Obviously just a taste thing, but it makes it so the condition is the place where the special case work happens, and the non-conditional flow is identical every time.